File: android/src/main/kotlin/in/agnostech/flutter_stripe_terminal/FlutterStripeTerminal.kt
package `in`.agnostech.flutter_stripe_terminal

import android.os.Handler
import android.os.Looper
import android.util.Log
import com.stripe.stripeterminal.Terminal
import com.stripe.stripeterminal.external.callable.Callback
import com.stripe.stripeterminal.external.callable.Cancelable
import com.stripe.stripeterminal.external.callable.PaymentIntentCallback
import com.stripe.stripeterminal.external.callable.ReaderCallback
import com.stripe.stripeterminal.external.models.*
import io.flutter.plugin.common.MethodChannel

class FlutterStripeTerminal {
    companion object {
        lateinit var serverUrl: String
        lateinit var authToken: String
        var availableReadersList: List<Reader>? = null
        var flutterStripeTerminalEventHandler: FlutterStripeTerminalEventHandler? = null
        var cancelDiscovery: Cancelable? = null;

        fun setConnectionTokenParams(
            serverUrl: String,
            authToken: String,
            result: MethodChannel.Result
        ) {
            this.serverUrl = serverUrl
            this.authToken = authToken
            result.success(true)
        }

        fun disconnectReader(result: MethodChannel.Result) {
            cancelDiscovery?.cancel(object: Callback {
                override fun onFailure(e: TerminalException) {
                    Handler(Looper.getMainLooper()).post {
                        result.error(e.errorCode.toLogString(), e.message, null)
                    }
                }

                override fun onSuccess() {
                    Handler(Looper.getMainLooper()).post {
                        Log.d("STRIPE TERMINAL", "reader discovery cancelled")
                    }
                }

            })
            Terminal.getInstance().disconnectReader(object: Callback {
                override fun onFailure(e: TerminalException) {
                    Handler(Looper.getMainLooper()).post {
                        result.error(e.errorCode.toLogString(), e.message, null)
                    }
                }

                override fun onSuccess() {
                    Handler(Looper.getMainLooper()).post {
                        result.success(true)
                    }
                }

            })
        }

        fun searchForReaders(result: MethodChannel.Result) {
            val config = DiscoveryConfiguration(
                discoveryMethod = DiscoveryMethod.BLUETOOTH_SCAN
            )
            cancelDiscovery = Terminal.getInstance().discoverReaders(
                config,
                flutterStripeTerminalEventHandler!!.getDiscoveryListener(),
                object : Callback {
                    override fun onSuccess() {
                        Handler(Looper.getMainLooper()).post {
                            result.success(true)
                        }
                    }

                    override fun onFailure(e: TerminalException) {
                        Handler(Looper.getMainLooper()).post {
                            result.error(e.errorCode.toLogString(), e.message, null)
                        }
                    }
                })
        }

        fun connectToReader(readerSerialNumber: String, locationId: String, result: MethodChannel.Result) {

            val reader = availableReadersList!!.filter {
                it.serialNumber == readerSerialNumber
            }

            if (reader.isNotEmpty()) {
                Terminal.getInstance().connectBluetoothReader(reader[0],
                    ConnectionConfiguration.BluetoothConnectionConfiguration(locationId),
                    flutterStripeTerminalEventHandler!!.getBluetoothReaderListener(),
                    object : ReaderCallback {
                        override fun onFailure(e: TerminalException) {
                            Handler(Looper.getMainLooper()).post {
                                result.error(e.errorCode.toLogString(), e.message, null)
                            }
                        }

                        override fun onSuccess(reader: Reader) {
                            Handler(Looper.getMainLooper()).post {
                                result.success(true)
                            }
                        }
                    })
            }
        }

        fun processPayment(clientSecret: String, result: MethodChannel.Result) {
            val terminal = Terminal.getInstance()
            terminal.retrievePaymentIntent(clientSecret, object: PaymentIntentCallback {
                override fun onFailure(e: TerminalException) {
                    Handler(Looper.getMainLooper()).post {
                        result.error(e.errorCode.toLogString(), e.message, null)
                    }
                }

                override fun onSuccess(paymentIntent: PaymentIntent) {
                    Log.d("STRIPE TERMINAL", "payment intent retrieved");
                    Handler(Looper.getMainLooper()).post {
                        terminal.collectPaymentMethod(paymentIntent, object: PaymentIntentCallback {
                            override fun onFailure(e: TerminalException) {
                                Handler(Looper.getMainLooper()).post {
                                    result.error(e.errorCode.toLogString(), e.message, null)
                                }
                            }

                            override fun onSuccess(paymentIntent: PaymentIntent) {
                                terminal.processPayment(paymentIntent, object: PaymentIntentCallback {
                                    override fun onFailure(e: TerminalException) {
                                        Handler(Looper.getMainLooper()).post {
                                            result.error(e.errorCode.toLogString(), e.message, null)
                                        }
                                    }

                                    override fun onSuccess(paymentIntent: PaymentIntent) {
                                        Handler(Looper.getMainLooper()).post {
                                            result.success(mapOf(
                                                "paymentIntentId" to paymentIntent.id
                                            ))
                                        }
                                    }

                                })
                            }

                        })
                    }
                }

            })
        }
    }
}

File: ios/Classes/FlutterStripeTerminal.swift
//
//  FlutterStripeTerminal.swift
//  flutter_stripe_terminal
//
//  Created by Vishal Dubey on 03/12/21.
//

import Foundation
import StripeTerminal
import Flutter
import Dispatch

class FlutterStripeTerminal {
    static let shared = FlutterStripeTerminal()
    
    var serverUrl: String?
    var authToken: String?
    var availableReaders: [Reader]?
    var discoverCancelable: Cancelable?
    
    func setConnectionTokenParams(serverUrl: String, authToken: String, result: FlutterResult) {
        self.serverUrl = serverUrl
        self.authToken = authToken
        Terminal.setTokenProvider(APIClient.shared)
        result(true)
    }
    
    func searchForReaders(result: @escaping FlutterResult) {
        let config = DiscoveryConfiguration(discoveryMethod: DiscoveryMethod.bluetoothScan, simulated: false)
        self.discoverCancelable = Terminal.shared.discoverReaders(config, delegate: FlutterStripeTerminalEventHandler.shared) { error in
            DispatchQueue.main.async {
                if let error = error {
                    result(FlutterError(code: "Search Error", message: "", details: ""))
                } else {
                    result(true)
                }
            }
        }
    }
    
    func connectToReader(readerSerialNumber: String, locationId: String, result: @escaping FlutterResult) {
        let selectedReaders = availableReaders?.filter { reader in
            return reader.serialNumber == readerSerialNumber
        }
       
        if (!selectedReaders!.isEmpty) {
            Terminal.shared.connectBluetoothReader(selectedReaders![0], delegate: FlutterStripeTerminalEventHandler.shared, connectionConfig: BluetoothConnectionConfiguration(locationId: locationId)) { _, error in
                DispatchQueue.main.async {
                    if let error = error {
                        result(FlutterError(code: "Connect Error", message: "", details: ""))
                    } else {
                        result(true)
                    }
                }
            }
        }
    }
    
    func processPayment(clientSecret: String, result: @escaping FlutterResult) {
        let terminal = Terminal.shared
        terminal.retrievePaymentIntent(clientSecret: clientSecret) { retrievedIntent, retrievedError in
            if let retrievedPaymentIntent = retrievedIntent {
                terminal.collectPaymentMethod(retrievedPaymentIntent) { processedIntent, processedError in
                    if let processedPaymentIntent = processedIntent {
                        terminal.processPayment(processedPaymentIntent) {finalIntent, finalError in
                            if let finalPaymentIntent = finalIntent {
                                result([
                                    "paymentIntentId": finalPaymentIntent.stripeId
                                ])
                            } else if let finalError = finalError {
                                DispatchQueue.main.async {
                                    result(FlutterError(code: "Final Intent Error", message: "", details: ""))
                                }
                            }
                        }
                    } else if let processedError = processedError {
                        DispatchQueue.main.async {
                            result(FlutterError(code: "Processed Error", message: "", details: ""))
                        }
                    }
                }
            } else if let retrievedError = retrievedError {
                DispatchQueue.main.async {
                    result(FlutterError(code: "Retrived Error", message: "", details: ""))
                }
            }
        }
    }
    
    func disconnectReader(result: @escaping FlutterResult) {
        if(self.discoverCancelable == nil){
            Terminal.shared.disconnectReader() { error in
                DispatchQueue.main.async {
                    if let error = error {
                        result(FlutterError(code: "Disconnect Reader Error", message: "", details: ""))
                    } else {
                        result(true)
                    }
                }
            }
                    
        }else {
            self.discoverCancelable?.cancel() { error in
                Terminal.shared.disconnectReader() { error in
                    DispatchQueue.main.async {
                    if let error = error {
                        result(FlutterError(code: "Disconnect Reader Error", message: "", details: ""))
                    } else {
                        result(true)
                    }
                }
                }
            }
            
        }
    }
}
